src/components/fields/text-field/text-field.tsx

```diff
@@ -91,12 +94,42 @@ export const TextField = (props: IGenericFieldProps<ITextFieldSchema | IEmailFie
 		// eslint-disable-next-line react-hooks/exhaustive-deps
 	}, [value]);
 
+	useEffect(() => {
+		if (props.schema.uiType === "text-field" && ref.current.selectionEnd !== caret.current) {
+			// keep caret in place after uppercase, not available for 'email' & 'number' HTML input types
+			ref.current.setSelectionRange(caret.current, caret.current);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [stateValue]);
+
+	useEffect(() => {
+		if (
+			props.schema.uiType === "text-field" &&
+			props.schema.customOptions?.textTransform === "uppercase" &&
+			stateValue
+		) {
+			setStateValue((previous) => (previous ? ("" + previous).toUpperCase() : ""));
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [props.schema.uiType === "text-field" && props.schema.customOptions?.textTransform]);
+
 	// =============================================================================
 	// EVENT HANDLERS
 	// =============================================================================
 	const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
 		if (uiType === "numeric-field" && !event.target.value) {
 			onChange({ target: { value: undefined } });
+		} else if (uiType === "text-field") {
+			const inputted = event.target.value;
+			const doUpperCase =
+				props.schema.uiType === "text-field" && props.schema.customOptions?.textTransform === "uppercase";
```

Comment:
- type inference will also work if you use the full path for the uiType in the if condition
- can refactor a bit to refer via the schema instead of the props

```suggestion
		} else if (schema.uiType === "text-field") {
			const inputted = event.target.value;
			const doUpperCase = schema.customOptions?.textTransform === "uppercase";
```

Reply:
interesting. thanks for sharing this. have used the above to simplify the code.